### PR TITLE
[vms/platformvm] Add tracking of a Subnet manager

### PR DESCRIPTION
**Pull request by dhrubabasu**
## Why this should be merged

[ACP-77](https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/77-reinventing-subnets/README.md) introduces a notion of a "Subnet manager" defined as a `sourceChainID` and `sourceAddress` pair. This PR adds tracking of this pair to the P-Chain state.

## How this works

Adds support for storing and retrieving a `chainID` and `addr` for each Subnet

## How this was tested

CI + added UTs

### CODE REVIEW

File: vms/platformvm/state/diff.go [hunk 1]
```diff
@@ -294,6 +296,27 @@ func (d *diff) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
 	d.subnetOwners[subnetID] = owner
 }
 
+func (d *diff) GetSubnetManager(subnetID ids.ID) (ids.ID, []byte, error) {
+	manager, exists := d.subnetManagers[subnetID]
+	if exists {
+		return manager.chainID, manager.addr, nil
+	}
+
+	// If the subnet owner was not assigned in this diff, ask the parent state.
```
Comment (dhrubabasu) [inline vms/platformvm/state/diff.go:305]:
```suggestion
	// If the subnet manager was not assigned in this diff, ask the parent state.
```

File: vms/platformvm/state/state.go [hunk 2]
```diff
@@ -409,6 +418,11 @@ type fxOwnerAndSize struct {
 	size  int
 }
 
+type chainIDAndAddr struct {
```
[the remark this answers is not in the document]
Comment (dhrubabasu) [reply vms/platformvm/state/state.go:421]: I thought about this when drafting the PR but decided to punt the decision till later in the ACP-77 development process. Right now, it's only used as an intermediate struct only for the codec marshalling. The PlatformVM `state` pkg is not expected for external use either so it wouldn't live here if we do export it.

Good call out!

File: vms/platformvm/state/state_test.go [hunk 3]
```diff
@@ -1401,6 +1401,28 @@ func TestStateSubnetOwner(t *testing.T) {
 	require.Equal(owner2, owner)
 }
 
+func TestStateSubnetManager(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dhrubabasu) [reply vms/platformvm/state/state_test.go:1404]: I think testing the database stuff is a bit difficult to do, it's been a long-standing TODO on the `state` pkg. Will punt for now just for the sake of progressing on ACP-77

File: vms/platformvm/config/execution_config.go [hunk 4]
```diff
@@ -36,6 +36,7 @@ type ExecutionConfig struct {
 	ChainDBCacheSize             int            `json:"chain-db-cache-size"`
 	BlockIDCacheSize             int            `json:"block-id-cache-size"`
 	FxOwnerCacheSize             int            `json:"fx-owner-cache-size"`
+	ChainIDAndAddrCacheSize      int            `json:"chain-id-and-addr-cache-size"`
```
Comment (StephenButtolph) [inline vms/platformvm/config/execution_config.go:39]: Do we not have a more concise name for this? `SubnetManager` or something? `ChainIDAndAddr` feels pretty amorphous.

File: vms/platformvm/state/diff.go [hunk 5]
```diff
@@ -294,6 +296,26 @@ func (d *diff) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
...
+	}
+	return parentState.GetSubnetManager(subnetID)
+}
+
+func (d *diff) SetSubnetManager(subnetID ids.ID, chainID ids.ID, addr []byte) {
+	d.subnetManagers[subnetID] = chainIDAndAddr{
+		ChainID: chainID,
+		Addr:    addr,
+	}
+}
```
Comment (StephenButtolph) [inline vms/platformvm/state/diff.go:317]: Don't we need to use `subnetManagers` during `Apply`? Can we add a test for this?

File: vms/platformvm/state/diff.go [hunk 6]
```diff
@@ -294,6 +296,26 @@ func (d *diff) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
...
+	// If the subnet manager was not assigned in this diff, ask the parent state.
+	parentState, ok := d.stateVersions.GetState(d.parentID)
+	if !ok {
+		return ids.Empty, nil, ErrMissingParentState
+	}
+	return parentState.GetSubnetManager(subnetID)
+}
+
+func (d *diff) SetSubnetManager(subnetID ids.ID, chainID ids.ID, addr []byte) {
+	d.subnetManagers[subnetID] = chainIDAndAddr{
```
Comment (StephenButtolph) [inline vms/platformvm/state/diff.go:313]: Where is `subnetManagers` initialized? Test?

File: vms/platformvm/state/state.go [hunk 7]
```diff
@@ -808,6 +837,42 @@ func (s *state) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
...
+		return chainIDAndAddr.ChainID, chainIDAndAddr.Addr, nil
+	}
+
+	if chainIDAndAddr, cached := s.subnetManagerCache.Get(subnetID); cached {
+		return chainIDAndAddr.ChainID, chainIDAndAddr.Addr, nil
+	}
+
+	chainIDAndAddrBytes, err := s.subnetManagerDB.Get(subnetID[:])
+	if err == database.ErrNotFound {
+		s.SetSubnetManager(subnetID, ids.Empty, []byte{})
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:851]: Why are we calling this here?

File: vms/platformvm/state/state.go [hunk 8]
```diff
@@ -808,6 +837,42 @@ func (s *state) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
...
+		return ids.Empty, []byte{}, nil
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:852]: Returning `nil` here seems unexpected to me... Why are these values returned?

File: vms/platformvm/state/state.go [hunk 9]
```diff
@@ -808,6 +837,42 @@ func (s *state) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
...
+	}
+
+	var manager chainIDAndAddr
+	if _, err := block.GenesisCodec.Unmarshal(chainIDAndAddrBytes, &manager); err != nil {
+		return ids.Empty, nil, err
+	}
+	s.subnetManagerCache.Put(subnetID, chainIDAndAddr{
+		ChainID: manager.ChainID,
+		Addr:    manager.Addr,
+	})
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:865]:
```suggestion
	s.subnetManagerCache.Put(subnetID, manager)
```

File: vms/platformvm/state/state.go [hunk 10]
```diff
@@ -2212,6 +2278,29 @@ func (s *state) writeSubnetOwners() error {
...
+
+		managerBytes, err := block.GenesisCodec.Marshal(block.CodecVersion, &manager)
+		if err != nil {
+			return fmt.Errorf("failed to marshal subnet manager: %w", err)
+		}
+
+		s.subnetManagerCache.Put(subnetID, chainIDAndAddr{
+			ChainID: manager.ChainID,
+			Addr:    manager.Addr,
+		})
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:2295]:
```suggestion
		s.subnetManagerCache.Put(subnetID, manager)
```

File: vms/platformvm/config/execution_config.go (hunk 4 above)
Comment (dhrubabasu) [reply vms/platformvm/config/execution_config.go:39]: was mostly taking the queue from `fx-owner-cache-size`, updated to `subnet-manager-cache-size`

File: vms/platformvm/state/diff.go (hunk 6 above)
Comment (dhrubabasu) [reply vms/platformvm/state/diff.go:313]: It wasn't 😓... fixed

File: vms/platformvm/state/diff.go (hunk 5 above)
Comment (dhrubabasu) [reply vms/platformvm/state/diff.go:318]: Yes, added a UT

File: vms/platformvm/state/state.go (hunk 7 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go:851]: My thought was to cache the `{ids.Empty, []byte{}}` as the subnet manager when we first do `GetSubnetManager` on a newly created Subnet

File: vms/platformvm/state/state.go (hunk 8 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go:852]: If we call `GetSubnetManager` on existing Subnets, we'll get an `ErrNotFound`. It's not that these are errors, it's that these Subnets don't have managers. Returning `{ids.Empty, []byte{}}` is the empty Subnet manager which is expected until `SetSubnetValidatorManagerTx` is called.

File: vms/platformvm/state/state.go (hunk 7 above)
Comment (StephenButtolph) [reply vms/platformvm/state/state.go:851]: This writes to the database though - no?

File: vms/platformvm/state/state.go (hunk 8 above)
Comment (StephenButtolph) [reply vms/platformvm/state/state.go:852]: I think that this is an unusual place to specify that "No manager is treated as `{ids.Empty, []byte{}}`" is defined in the state package rather than the execution logic.

Comment (dhrubabasu) [reply vms/platformvm/state/state.go]: Fair- will handle in a followup PR for the execution logic

File: vms/platformvm/state/state_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshua-kim) [reply vms/platformvm/state/state_test.go:1404]: i agree w/ @marun 's comment - it should be trivial to add test tables for this new test we're adding (we don't have to do this on the whole package). Also I don't think we're testing diff application + whether or not the writes are actually occurring to state + the missing parent state case

[the remark this answers is not in the document]
Comment (dhrubabasu) [reply vms/platformvm/state/state_test.go:1404]: The diff application is tested in diff_test.go (`TestDiffSubnetManager`)

File: vms/platformvm/state/state.go [hunk 11]
```diff
@@ -339,6 +343,10 @@ type state struct {
 	subnetOwnerCache cache.Cacher[ids.ID, fxOwnerAndSize] // cache of subnetID -> owner; if the entry is nil, it is not in the database
 	subnetOwnerDB    database.Database
 
+	subnetManagers     map[ids.ID]chainIDAndAddr            // map of subnetID -> manager of the subnet
+	subnetManagerCache cache.Cacher[ids.ID, chainIDAndAddr] // cache of subnetID -> manager; if the entry is nil, it is not in the database
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:347]: The entry can't be `nil` as it is a struct - no?

File: vms/platformvm/state/state.go [hunk 12]
```diff
@@ -547,6 +560,18 @@ func newState(
 		return nil, err
 	}
 
+	subnetManagerDB := prefixdb.New(SubnetManagerPrefix, baseDB)
+	subnetManagerCache, err := metercacher.New[ids.ID, chainIDAndAddr](
+		"subnet_manager_cache",
+		metricsReg,
+		cache.NewSizedLRU[ids.ID, chainIDAndAddr](execCfg.SubnetManagerCacheSize, func(_ ids.ID, f chainIDAndAddr) int {
+			return ids.IDLen + len(f.Addr)
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:568]: We typically account for both the key size and the value size.
```suggestion
			return 2*ids.IDLen + len(f.Addr)
```

File: vms/platformvm/state/state.go (hunk 11 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go]: Good catch, fixed the comment

File: vms/platformvm/state/state.go [hunk 13]
```diff
@@ -339,6 +343,10 @@ type state struct {
 	subnetOwnerCache cache.Cacher[ids.ID, fxOwnerAndSize] // cache of subnetID -> owner; if the entry is nil, it is not in the database
 	subnetOwnerDB    database.Database
 
+	subnetManagers     map[ids.ID]chainIDAndAddr            // map of subnetID -> manager of the subnet
+	subnetManagerCache cache.Cacher[ids.ID, chainIDAndAddr] // cache of subnetID -> manager; if the entry is {ids.Empty, nil}, it is not in the database
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:347]: We don't cache the entry if it is not in the database.

File: vms/platformvm/state/state_test.go (hunk 3 above)
Comment (StephenButtolph) [inline vms/platformvm/state/state_test.go:1404]: 1. One thing we could test here is the cache usage.
2. Another interesting edge case to consider is: "what is the maximum Address size?"

Comment (dhrubabasu) [reply vms/platformvm/state/state_test.go:1404]: Added a cache test.

For the maximum address size, I think bounding it at `4096` is more than enough. [ML-DSA-87](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.204.ipd.pdf) (previously known as DILITHIUM) has a public key size of 2592. I don't think this should be bounded on the `state` interface and should instead be bounded on the `ConvertSubnetTx` transaction.

File: vms/platformvm/state/diff_test.go [hunk 14]
```diff
@@ -580,6 +580,55 @@ func TestDiffSubnetOwner(t *testing.T) {
 	require.Equal(owner2, owner)
 }
 
+func TestDiffSubnetManager(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dhrubabasu) [reply vms/platformvm/state/diff_test.go:583]: This test is testing one thing: is the diff and state correct before the diff is applied and after

table tests can be used for the state test. did it here: cda1001

File: vms/platformvm/state/state.go [hunk 15]
```diff
@@ -338,6 +342,10 @@ type state struct {
 	subnetOwnerCache cache.Cacher[ids.ID, fxOwnerAndSize] // cache of subnetID -> owner; if the entry is nil, it is not in the database
 	subnetOwnerDB    database.Database
 
+	subnetManagers     map[ids.ID]chainIDAndAddr            // map of subnetID -> manager of the subnet
+	subnetManagerCache cache.Cacher[ids.ID, chainIDAndAddr] // cache of subnetID -> manager
```
[the remark this answers is not in the document]
Comment (dhrubabasu) [reply vms/platformvm/state/state.go:346]: The `map` is populated with any new subnet managers that are added during node operation. At shutdown, this in-memory map is flushed to the database. At startup, the `map` is not populated with all the database entries.

This is where the cache comes in. If `GetSubnetManager` is called, it does a db read. If it's a hit, it is stored in the `subnetManagerCache` to hopefully save on db reads in the future.

File: vms/platformvm/config/execution_config.go [hunk 16]
```diff
@@ -21,6 +21,7 @@ var DefaultExecutionConfig = ExecutionConfig{
 	ChainDBCacheSize:             2048,
 	BlockIDCacheSize:             8192,
 	FxOwnerCacheSize:             4 * units.MiB,
+	SubnetManagerCacheSize:       2048,
```
Comment (StephenButtolph) [inline vms/platformvm/config/execution_config.go:24]: This is pretty confusing. All of the bytes caches use `units.MiB` to be clear that it is representing bytes. All of the other values are denominating the number of entries.

This entry is interpreted as bytes - but is set as if it is in the number of entries.

I think we should either be specifying `2 * units.KiB` or we should be setting this to a different value (if the assumption was that this is denominated in entries rather than bytes)

File: vms/platformvm/state/state_test.go [hunk 17]
```diff
@@ -1401,6 +1401,56 @@ func TestStateSubnetOwner(t *testing.T) {
...
+	}{
+		{
+			name: "in-memory",
+			setup: func(s State, subnetID ids.ID, chainID ids.ID, addr []byte) {
+				s.SetSubnetManager(subnetID, chainID, addr)
+			},
+		},
+		{
+			name: "cache",
+			setup: func(s State, subnetID ids.ID, chainID ids.ID, addr []byte) {
```
Comment (StephenButtolph) [inline vms/platformvm/state/state_test.go:1417]: Should this pass in the `*testing.T` or the `*require.Assertions`? Right now this function is executed by the inner test but uses the outer testing struct.

**Review by StephenButtolph — APPROVED** [review]
